dotCMS/src/main/java/com/dotcms/security/apps/AppsCacheImpl.java

```diff
@@ -0,0 +1,213 @@
+        if (!UtilMethods.isSet(appDescriptors) && null != supplier) {
+           appDescriptors = supplier.get();
+        }
+        putAppDescriptor(appDescriptors);
+        return appDescriptors;
+    }
+
+    /**
+     * This will simply put a list of descriptor into memory.
+     * @param appDescriptors
+     */
+    private void putAppDescriptor(final List<AppDescriptorMeta> appDescriptors) {
+        cache.put(DESCRIPTORS_LIST_KEY, appDescriptors, DESCRIPTORS_CACHE_GROUP);
+    }
+
+    /**
+     * Given a supplier delegate able to retrieve AppDescriptors this method will use it to populate cache.
+     * Once passed that point the method takes whatever descriptors exist in cache and use the returned entries
+     * to build Map like structure still in cache to track AppDescriptors by app-key.
+     * If the supplier is null the function will get back the contents from cache and no population attempt will be performed.
+     * @param supplier
+     * @return
+     */
+    public Map<String, AppDescriptorMeta> getAppDescriptorsMap(final Supplier<List<AppDescriptorMeta>> supplier) {
+        Map<String, AppDescriptorMeta> descriptorsByKey = (Map<String, AppDescriptorMeta>) cache.getNoThrow(
+                DESCRIPTORS_MAPPED_BY_KEY, DESCRIPTORS_CACHE_GROUP);
+        if (!UtilMethods.isSet(descriptorsByKey) && null != supplier) {
+            synchronized (AppsCacheImpl.class) {
+                descriptorsByKey = getAppDescriptorsMeta(supplier).stream().collect(
+                        Collectors.toMap(serviceDescriptorMeta -> serviceDescriptorMeta
+                                        .getAppDescriptor().getKey().toLowerCase(), Function.identity(),
+                                (serviceDescriptor, serviceDescriptor2) -> serviceDescriptor));
+
+                putDescriptorsByKey(descriptorsByKey);
+            }
+        }
+        return descriptorsByKey;
+    }
+
+    /**
+     * This simply allow putting in cache
+     * A Map like structure, Where the key is the app-key and the entry is the AppDescriptor.
+     * @param descriptorsByKey
+     */
+    private void putDescriptorsByKey(final Map<String, AppDescriptorMeta> descriptorsByKey){
+        cache.put(DESCRIPTORS_MAPPED_BY_KEY, descriptorsByKey, DESCRIPTORS_CACHE_GROUP);
+    }
+
+    /**
+     * Remove all descriptors from cache;
+     */
+    public void invalidateDescriptorsCache(){
+        cache.flushGroup(DESCRIPTORS_CACHE_GROUP);
+    }
+
+    /**
+     * In a similar fashion this takes a supplier delegate responsible for getting the secret keys to populate cache.
+     * If the supplier is null the function will get back the contents from cache and no population attempt will be performed.
+     * @param supplier string keys set delegate
+     * @return
+     * @throws DotCacheException
+     */
+    public synchronized Set<String> getKeysFromCache(final Supplier<Set<String>> supplier) throws DotCacheException {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use block level rather than method level synchronization](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5358201)